### DISCUSSION TITLE
User password expiry -hidden feature?

### DISCUSSION THREAD

**Discussion by PeterBowater** [MasaCMS/MasaCMS#55] [Ideas]
Hi all!

Was digging around in the Masa code, and came across something that I thought would be useful, so pursued it a bit.....

In May 2020, a [settings.ini.cfm entry was added](https://github.com/MasaCMS/MasaCMS/commit/611e62e070657d8ffd58f8a334300e52b2399074) enabling password expiry until today I was unaware of this as a feature and had even explained to someone that had asked the question tat it probably wasn't possible!

But apparently it is, but using these two keys in settings.ini.cfm we can set the password expiry policy:

passwordsExpire=true
expirePasswords=90

(90 = 90 days!)

This appears to work pretty well, though nothing is actually enforced when this is enabled, there is a red banner at the top of every page you visit in the CMS that tells you to change it:

![Screenshot 2022-03-01 at 16 01 37](https://user-images.githubusercontent.com/807991/156209138-74b06e0b-b582-4445-9db2-c4216b7382e8.jpg)

This is permanent, so even if you dismiss it, on the next page it reappears, so is quite a good nag :) 

Additionally, when you edit a user there is a[ button](https://github.com/MasaCMS/MasaCMS/blob/ea9dd65d88ba55d5b41131cf355778a6529ff42a/admin/core/views/cusers/edituser.cfm#L226) which allows you to expire that user's password immediately -unfortunately that fails with a 500 message saying that the method does not exist, so assume this work got left unfinished.... There are two keys that need adding en.properties (or your chosen lang...) to fill in the button and dialog box blanks:

user.expirepassword=Expire password
user.expirepasswordconfirm=Are you sure you want to expire this user's password?

Then you get this:

![Screenshot 2022-03-01 at 16 39 28](https://user-images.githubusercontent.com/807991/156210543-be22171a-6d3a-484c-9fec-d16bf1b243d1.jpg)

![Screenshot 2022-03-01 at 16 39 35](https://user-images.githubusercontent.com/807991/156210551-10a18005-34fc-488c-959d-c32e302f7c0e.jpg)

Even with that, it is probably almost good enough for most situations, but just thinking around how it could be improved I came up with some suggestions, which may be rubbish, but this is a discussion forum, right? ;) 

- Rather than (or maybe as well as) the red banner, forward the user to their profile page so they are forced to change their password before doing anything else (could be on another switch if necessary)
- Fix the 'Expire password' button so it does what it says
- anything else?

For something that seems to close, it would be nice to get it fully featured 💯 

Cheers,

Peter

**Comment by guustnieuwenhuis** [comment]
Hi @PeterBowater, Thanks a lot for this post!

Could you add the settings to the documentation? https://docs.masacms.com/getting-started/configuration/configuration-file/
And create issues for the improvements? Security is high priority for us, so we'll pick them up in a future release!
